### DISCUSSION TITLE
Musl support status

### DISCUSSION THREAD

**Discussion by yveszoundi** [fltk-rs/fltk-rs#1354] [General]
### Actual Question

What is the current stance regarding `MUSL` support for `FLTK-RS`? (i.e. try and see, not available, etc.)

### Background 

I recently built a new Linux virtual machine (Void Linux x86_64-musl) and I realized that my applications don't run or compile.

- I haven't attempted to compile on Alpine yet, I'll probably test under Docker/Podman. I run into different errors on Ubuntu with the `x86_64-unknown-linux-musl` rust target.
- At first sight it feels like that `MUSL` support would be more of an upstream `FLTK` issue, rather than something related to the `CFLTK` build system. 

I see more people using MUSL systems  outside embedded environments or containers (Alpine, Void Linux MUSL edition, Gentoo MUSL build, etc.). Nowadays, I don't know if users would feel "deceived" if developers don't state explicitly "only for GLIBC".

### Additional Information

This is a screenshot of a build attempt 
- "-crt-static" set, `fltk-shared` feature instead of `fltk-bundled` but I tried both 
- The usual dev/devel libraries are installed (xkb, X11 and/or wayland, pango, etc.).

<img width="1440" alt="test12" src="https://user-images.githubusercontent.com/5206067/218191091-725da4d8-0cd0-4f50-bcc2-29fa1113f351.png">

### References

- https://github.com/fltk-rs/fltk-rs/issues/1228
- https://github.com/fltk-rs/fltk-rs/issues/263

**Comment by MoAlyousef** [comment]
You should be able to build for musl on alpine linux on wsl and or on a vm. I’ve also tested with docker (see [here](https://fltk-rs.github.io/fltk-book/Cross-Compiling.html#using-docker)) and cross-rs. The idea is that the system itself has to be musl based. If it tries to link to xlib which was built with glibc, it will fail, since on linux, fltk requires some system deps and these have to be built with musl as well. Also if you use fltk-bundled, it will fail since we currently don’t provide a prebuilt bundle for musl.

**Reply by yveszoundi** [reply]
Thanks, I'll try on Alpine instead of Void Linux shortly. I figured so for the usage of `fltk-shared` which is what I'm using in the previous screenshot.

**Reply by MoAlyousef** [reply]
I'm not sure if fltk-shared would work or not. I think it should generally speaking. It should output a libcfltk.so file which would need to be deployed with your app.

**Reply by yveszoundi** [reply]
Thanks, just confirming that  it works well and I was able to build the UI component inside Alpine and launch it in Void Linux

- I cloned the `fltk-rs` repository and pulled its submodules
- I updated the dependency reference inside my `cargo` manifest  `fltk = { path = "../../fltk-rs/fltk"  }`
- Then I was able to compile successfully without any special flag
- To run it doesn't appear that I need any specific `libcfltk.so`, just standard dependencies such as xcb among other things. You also guessed right for `fltk-shared`, it's a no go from very quick tests.

I repeated the above steps directly on Void Linux musl (instead of Podman+Alpine) with the same successful results.

**Reply by MoAlyousef** [reply]
You should be able to build directly without having to clone fltk-rs separately. You would just need to remove the fltk-bundled flag from your app’s Cargo.toml. You can create a cfg target dependency so it treats musl based builds differently. 

```toml
[target.'cfg(target_env = "musl")'.dependencies]
fltk = "1.3.30"
```

Or you can remove the fkag altogether and just invoke it from the command line on platforms that have bundles:
`cargo build —features=fltk/fltk-bundled`

**Reply by yveszoundi** [reply]
Thank you
